vllm/attention/ops/blocksparse_attention/blocksparse_attention_kernel.py

```diff
@@ -3,6 +3,407 @@
+        other=0.0)
+
+    # layout_crow_ptr shape is [num_heads, num_blocks + 1], where
+    # num_blocks = max_seqlen // block_size
+    sparse_crow_ptr = (layout_crow_ptr + cur_head * layout_crow_stride_h +
+                       block_id_for_current_query_block * layout_crow_stride_m)
+
+    # Find the k block that current query block should attend to.
+    # inclusive
+    k_block_start = tl.load(sparse_crow_ptr).to(tl.int32)
+    # exclusive
+    k_block_end = tl.load(sparse_crow_ptr + 1).to(tl.int32)
+
+    # on-chip memory
+    # running max value of attention score
+    m_i = tl.zeros([sparse_block_size], dtype=tl.float32) - float("inf")
+    # running sum of EXP of attention score
+    l_i = tl.zeros([sparse_block_size], dtype=tl.float32)
+    # The output, it won't be the final value until we process all the k blocks
+    acc = tl.zeros([sparse_block_size, BLOCK_D], dtype=tl.float32)
+
+    # initialize offsets
+    # [N]; starts at 0
+    offs_n = tl.arange(0, sparse_block_size)
+
+    sm_scale *= (
+        1.44269504  # 1/log2 as we use base2 for exponential and logarithm
+    )
+    # flash attention style tiling
+    for k_block_col_idx in range(k_block_start, k_block_end):
+        # The block id of the key matrix that the current query block should
+        # attend to.
+        k_block_id = tl.load(layout_col_ptr + cur_head * layout_col_stride_h +
+                             k_block_col_idx * layout_col_stride_m).to(
+                                 tl.int32)
+        # attend to k_cache blocks
+        if k_block_id <= context_end_block_id:
+            # -- compute qk ----
+            start_n = k_block_id * sparse_block_size
+            # load the token block ids for the current key matrix
+            bn = tl.load(
+                block_tables +
+                batch_id_for_current_query_block * stride_b_loc_b +
+                ((start_n + offs_n) // kv_cache_block_size) * stride_b_loc_s,
+                mask=(start_n + offs_n) < context_len,
+                other=0)  # [N]
+            # [D,N]
+            # Index into the key matrix cache
+            off_k = (bn[None, :] * stride_k_cache_bs +
+                     cur_kv_head * stride_k_cache_h +
+                     (offs_d[:, None] // x) * stride_k_cache_d +
+                     ((start_n + offs_n[None, :]) % kv_cache_block_size) *
+                     stride_k_cache_bl +
+                     (offs_d[:, None] % x) * stride_k_cache_x)
+            # [N,D]
+            off_v = (bn[:, None] * stride_v_cache_bs +
+                     cur_kv_head * stride_v_cache_h +
+                     offs_d[None, :] * stride_v_cache_d +
+                     (start_n + offs_n[:, None]) % kv_cache_block_size *
+                     stride_v_cache_bl)
+            k = tl.load(K_cache + off_k,
+                        mask=(start_n + offs_n[None, :]) < context_len,
+                        other=0.0)  # [D,N]
```

Comment:
based on a lack of masking along N here, looks like we require that the sparse block size divides the hidden dim? I think we should assert that somewhere if we don't already. I acknowledge this is paranoia.